unison-cli/src/Unison/Share/Sync.hs

```diff
@@ -606,64 +619,141 @@ getCausalHashByPath httpClient unisonShareUrl repoPath =
+          Share.UploadEntitiesNeedDependencies (Share.NeedDependencies moreHashes) -> Right (NESet.toSet moreHashes)
+          Share.UploadEntitiesNoWritePermission _ -> Left ()
+          Share.UploadEntitiesHashMismatchForEntity _ -> error "hash mismatch; fixme"
+          Share.UploadEntitiesSuccess -> Right Set.empty
+
+      case result of
+        Left () -> void (atomically (tryPutTMVar workerFailedVar ()))
+        Right moreHashes -> do
+          uploaded callbacks (NESet.size hashes)
+          (maybeYoungestWorkerThatWasAlive, numNewHashes) <-
+            atomically do
+              -- Record ourselves as "dead". The only work we have left to do is remove the hashes we just uploaded from
+              -- the `dedupe` set, but whether or not we are "alive" is relevant only to:
+              --
+              --   - The main dispatcher thread, which terminates when there are no more hashes to upload, and no alive
+              --     workers. It is not important for us to delete from the `dedupe` set in this case.
+              --
+              --   - Other worker threads, each of which independently decides when it is safe to delete the set of
+              --     hashes they just uploaded from the `dedupe` set (as we are doing now).
+              !workers <- Set.delete workerId <$> readTVar workersVar
+              writeTVar workersVar workers
+              -- Add more work (i.e. hashes to upload) to the work queue (really a work set), per the response we just
+              -- got from the server. Remember to only add hashes that aren't in the `dedupe` set (see the comment on
+              -- the dedupe set above for more info).
+              numNewHashes <-
+                if not (Set.null moreHashes)
+                  then do
+                    dedupe <- readTVar dedupeVar
+                    hashes0 <- readTVar hashesVar
+                    let !hashes1 = Set.union (Set.difference moreHashes dedupe) hashes0
+                    writeTVar hashesVar hashes1
+                    pure (Set.size hashes1 - Set.size hashes0)
+                  else pure 0
+              pure (Set.lookupMax workers, numNewHashes)
+          toUpload callbacks numNewHashes
+          -- Block until we are sure that the server does not have any uncommitted transactions that see a version of
+          -- the database that does not include the entities we just uploaded. After that point, it's fine to remove the
+          -- hashes of the entities we just uploaded from the `dedupe` set, because they will never be relevant for any
+          -- subsequent deduping operations. If we didn't delete from the `dedupe` set, this algorithm would still be
+          -- correct, it would just use an unbounded amount of memory to remember all the hashes we've uploaded so far.
+          whenJust maybeYoungestWorkerThatWasAlive \youngestWorkerThatWasAlive -> do
```

Comment:
Hrmm, I understand what this is doing; but it's very complex, and even though we have all the necessary information in the client, it kind of seems like an implementation detail of the server.

I suppose this is something else that can go away when we move to streaming right?

Reply:
Yeah, this can go away when we move to streaming, because the server won't have multiple simultaneous transactions open that all correspond to our push.